File: 2021/elixir/Day0104.ex
#
# Author: Loading BG#7962
#

defmodule Day0104 do
  defp split([_]),  do: []
  defp split([x, y]), do: [[x, y]]
  defp split([x, y | rest]), do: [[x, y]] ++ split([y] ++ rest)
  
  def check(nums) do
    sum = split(nums)
            |> Enum.map(fn [f, s | _] -> f - s end)
            |> Enum.sum
    cond do
      sum == length(nums) - 1    -> "decreasing"
      sum == -(length(nums) - 1) -> "increasing"
      :else                      -> "neither"
    end
  end
  
  defp area([x1, y1], [x2, y2], [x3, y3]), do: abs((x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)) / 2)
  def within_tri?([p1, p2, p3], p), do: area(p1, p2, p3) == area(p1, p2, p) + area(p1, p3, p) + area(p2, p3, p)
  
  defp color(<<l::utf8, n::utf8>>), do: rem(l, 2) == rem(n, 2)
  def bishop(start, finish, _), do: color(start) == color(finish)
end

IO.inspect Day0104.check([1, 2, 3]), label: "check #1"
IO.inspect Day0104.check([3, 2, 1]), label: "check #2"
IO.inspect Day0104.check([1, 2, 1]), label: "check #3"
IO.puts "---------"
IO.inspect Day0104.within_tri?([[1, 4], [5, 6], [6, 1]], [4, 5]), label: "withinTriangle #1"
IO.inspect Day0104.within_tri?([[1, 4], [5, 6], [6, 1]], [3, 2]), label: "withinTriangle #2"
IO.puts "---------"
IO.inspect Day0104.bishop("a1", "b4", 2), label: "bishop #1"
IO.inspect Day0104.bishop("a1", "b5", 5), label: "bishop #2"
IO.inspect Day0104.bishop("f1", "f1", 0), label: "bishop #3"

File: 2021/java/Day0104.java
/**
 * Author: Loading BG#7962
 */

import java.awt.Polygon;
import java.util.ArrayList;
import java.util.List;

class Day0104 {
    public static void main(String[] args) {
        System.out.println("check #1: " + check(new int[] {1, 2, 3}));
        System.out.println("check #2: " + check(new int[] {3, 2, 1}));
        System.out.println("check #3: " + check(new int[] {1, 2, 1}));
        System.out.println("----------");
        System.out.println("withinTriangle #1: " + withinTriangle(new int[][] {{1, 4}, {5, 6}, {6, 1}}, new int[] {4, 5}));
        System.out.println("withinTriangle #2: " + withinTriangle(new int[][] {{1, 4}, {5, 6}, {6, 1}}, new int[] {3, 2}));
        System.out.println("----------");
        System.out.println("bishop #1: " + bishop("a1", "b4", 2));
        System.out.println("bishop #2: " + bishop("a1", "b5", 5));
        System.out.println("bishop #3: " + bishop("f1", "f1", 0));
    }
    
    private static String check(int[] nums) {
        for (int i = 0; i < nums.length; ++i) {
            if (i == nums.length - 1) {
                return "increasing";
            }
            if (nums[i] > nums[i + 1]) {
                break;
            }
        }
        
        for (int i = 0; i < nums.length; ++i) {
            if (i == nums.length - 1) {
                return "decreasing";
            }
            if (nums[i] < nums[i + 1]) {
                return "neither";
            }
        }
        return null;
    }
    
    private static boolean withinTriangle(int[][] v, int[] point) {
        return new Polygon(new int[] {v[0][0], v[1][0], v[2][0]}, new int[] {v[0][1], v[1][1], v[2][1]}, 3)
            .contains(point[0], point[1]);
    }
    
    private static String[] possibleMoves(String pos) {
        List<String> moves = new ArrayList<>();
        
        char currPosL = pos.charAt(0);
        char currPosN = pos.charAt(1);
        while (currPosL >= 'a' && currPosN >= '1') {
            moves.add(new String(new char[] {currPosL, currPosN}));
            --currPosL;
            --currPosN;
        }
        
        currPosL = pos.charAt(0);
        currPosN = pos.charAt(1);
        while (currPosL >= 'a' && currPosN <= '8') {
            moves.add(new String(new char[] {currPosL, currPosN}));
            --currPosL;
            ++currPosN;
        }
        
        currPosL = pos.charAt(0);
        currPosN = pos.charAt(1);
        while (currPosL <= 'h' && currPosN <= '8') {
            moves.add(new String(new char[] {currPosL, currPosN}));
            ++currPosL;
            ++currPosN;
        }
        
        currPosL = pos.charAt(0);
        currPosN = pos.charAt(1);
        while (currPosL <= 'h' && currPosN >= '1') {
            moves.add(new String(new char[] {currPosL, currPosN}));
            ++currPosL;
            --currPosN;
        }
        
        return moves.toArray(new String[0]);
    }
    private static boolean bishop(String start, String end, int moves) {
        if (moves < 0) {
            return false;
        }
        if (start.equals(end)) {
            return true;
        }
        for (String move : possibleMoves(start)) {
            if (bishop(move, end, moves - 1)) {
                return true;
            }
        }
        return false;
    }
}
